toolchest_client/files.py

```diff
@@ -62,3 +71,120 @@ def files_in_path(files):
+
+    :param input_file_path: Path to the file which is to be split.
+    :param num_lines_in_group: Number of contiguous lines which cannot be split from one another.
+    :param max_bytes: Maximum size of each new file.
+    """
+    file_extension = pathlib.Path(input_file_path).suffix
+    if file_extension not in [".fastq", ".fasta", ".fa", ".fq", ".fna"]:
+        raise ValueError("Cannot split a non FASTQ/FASTA file for parallelization")
+
+    input_basename = os.path.basename(input_file_path)
+    current_split_number = 0
+    current_output_bytes = 0
+    current_line_number = 0
+    current_output_file_path, current_output_file = open_new_output_file(
+        current_split_number=current_split_number,
+        input_basename=input_basename
+    )
+    large_input_file = open(input_file_path, "r")
+
+    for line in large_input_file:
+        current_line_number += 1
+        # Assume that each character is one byte (inaccurate, but good enough for our use case)
+        bytes_in_line = len(line)
+        current_output_bytes += bytes_in_line
+        if current_output_bytes >= max_bytes and current_line_number % num_lines_in_group == 0:
+            # Time to switch output files
+            current_output_bytes = 0
+            current_output_file.close()
+            yield current_output_file_path
+            current_split_number += 1
+            current_output_file_path, current_output_file = open_new_output_file(
+                current_split_number=current_split_number,
+                input_basename=input_basename
+            )
+        else:
+            # Not time to switch output files yet
+            current_output_file.write(f"{line}")
+
+    current_output_file.close()
+    large_input_file.close()
+
+
+def sanity_check(file_path):
+    """Ensures file is greater than an arbitrary small size (5 bytes).
+
+    :param file_path: Path to the file which is to be checked.
+    """
+    assert_exists(file_path, must_be_file=True)
+    if os.stat(file_path).st_size <= 5:
+        raise ValueError(f"File at {file_path} is suspiciously small")
+
+
+def concatenate_files(input_file_paths, output_file_path):
+    """Concatenates a list of files using shutil.
+
+    :param input_file_paths: Paths to the files which are to be concatenated.
+    :param output_file_path: Path to the merged output file.
+    """
+    with open(output_file_path, "wb") as output_file:
+        for input_file_path in input_file_paths:
+            input_file = open(input_file_path, "rb")
+            shutil.copyfileobj(input_file, output_file)
+            input_file.close()
+
+
+def merge_sam_files(input_file_paths, output_file_path):
+    """Merges SAM files – the output for tools like STAR – using samtools.
+
+    :param input_file_paths: Paths to the files which are to be merged with samtools.
+    :param output_file_path: Path to the merged output file.
+    """
+    # This cause problems if run on a shared machine with non-available cores
```

Comment:
It might, but it shouldn't significantly impact performance.

For example, running this on a shared workspace (e.g. Jupyter) in which you are only allowed to utilize a certain number of cores (e.g. 2) when there are more actually present on the instance (e.g. 16).